tests/api/test_metrics_steps.py

```diff
@@ -0,0 +1,407 @@
+            raise RuntimeError()
+    except RuntimeError:
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.ERROR_KEY] == "RuntimeError"
+
+
+def test_metrics_steps_no_parent():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("step1"):
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.PARENT_KEY] is None
+
+
+def test_metrics_steps_with_parent_proper_names():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("parent"):
+        with metrics.track_step("child"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+
+    assert child[_CLIMetricsStep.NAME_KEY] == "child"
+    assert child[_CLIMetricsStep.PARENT_KEY] == "parent"
+    assert parent[_CLIMetricsStep.NAME_KEY] == "parent"
+
+
+def test_metrics_steps_with_parent_proper_ids():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("parent"):
+        with metrics.track_step("child"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+    assert child[_CLIMetricsStep.ID_KEY] != parent[_CLIMetricsStep.ID_KEY]
+    assert child[_CLIMetricsStep.PARENT_ID_KEY] == parent[_CLIMetricsStep.ID_KEY]
+    assert parent[_CLIMetricsStep.PARENT_ID_KEY] is None
+
+
+def test_metrics_steps_with_duplicate_names_proper_timing():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("duplicate"):
+        with metrics.track_step("duplicate"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+
+    assert (
+        child[_CLIMetricsStep.START_TIME_KEY] > parent[_CLIMetricsStep.START_TIME_KEY]
+    )
+    assert (
+        parent[_CLIMetricsStep.EXECUTION_TIME_KEY]
+        > child[_CLIMetricsStep.EXECUTION_TIME_KEY]
+    )
+
+
+def test_metrics_steps_error_caught_in_outer_step():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("parent"):
+        try:
+            with metrics.track_step("child"):
+                raise RuntimeError()
+        except RuntimeError:
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+    assert parent[_CLIMetricsStep.ERROR_KEY] is None
+    assert child[_CLIMetricsStep.ERROR_KEY] == "RuntimeError"
+
+
+def test_metrics_steps_manual_start_and_end_overlapping_proper_parent():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    step1_id = metrics.start_step("step1")
+    step2_id = metrics.start_step("step2")
+
+    metrics.end_step(step_id=step2_id)
+    metrics.end_step(step_id=step1_id)
+
+    # then
+    assert len(metrics.steps) == 2
+    step1, step2 = metrics.steps
+
+    assert step2[_CLIMetricsStep.PARENT_KEY] == "step1"
```

Comment:
yikes, that's surprising. Steps are implicitly nested? Let's discuss offline.

Reply:
Discussed offline: no more manual start/stop for spans in initial version of API, only supported through context managers, so spans are explicitly nested now